graph/bolt2/quadstore.go

```diff
@@ -0,0 +1,326 @@
+	qs.initBloomFilter()
+	return &qs, nil
+}
+
+func setVersion(db *bolt.DB, version int64) error {
+	return db.Update(func(tx *bolt.Tx) error {
+		buf := new(bytes.Buffer)
+		err := binary.Write(buf, binary.LittleEndian, version)
+		if err != nil {
+			clog.Errorf("Couldn't convert version!")
+			return err
+		}
+		b := tx.Bucket(metaBucket)
+		werr := b.Put([]byte("version"), buf.Bytes())
+		if werr != nil {
+			clog.Errorf("Couldn't write version!")
+			return werr
+		}
+		return nil
+	})
+}
+
+func (qs *QuadStore) Size() int64 {
+	return qs.size
+}
+
+func (qs *QuadStore) Close() error {
+	qs.db.Update(func(tx *bolt.Tx) error {
+		return qs.writeHorizonAndSize(tx)
+	})
+	err := qs.db.Close()
+	qs.open = false
+	return err
+}
+
+func (qs *QuadStore) getMetadata() error {
+	err := qs.db.View(func(tx *bolt.Tx) error {
+		var err error
+		qs.size, err = getInt64ForMetaKey(tx, "size", 0)
+		if err != nil {
+			return err
+		}
+		qs.version, err = getInt64ForMetaKey(tx, "version", nilDataVersion)
+		if err != nil {
+			return err
+		}
+		qs.sameAsHorizon, err = getInt64ForMetaKey(tx, "sameAsHorizon", 0)
+		if err != nil {
+			return err
+		}
+		qs.horizon, err = getInt64ForMetaKey(tx, "horizon", 0)
+		return err
+	})
+	return err
+}
+
+func getInt64ForMetaKey(tx *bolt.Tx, key string, empty int64) (int64, error) {
+	return getInt64ForKey(tx, metaBucket, key, empty)
+}
+
+func getInt64ForKey(tx *bolt.Tx, bucket []byte, key string, empty int64) (int64, error) {
+	var out int64
+	b := tx.Bucket(bucket)
+	if b == nil {
+		return empty, errNoBucket
+	}
+	data := b.Get([]byte(key))
+	if data == nil {
+		return empty, nil
+	}
+	buf := bytes.NewBuffer(data)
+	err := binary.Read(buf, binary.LittleEndian, &out)
+	if err != nil {
+		return 0, err
+	}
+	return out, nil
+}
+
+func (qs *QuadStore) Horizon() graph.PrimaryKey {
+	return graph.NewSequentialKey(qs.horizon)
+}
+
+func (qs *QuadStore) NameOf(k graph.Value) quad.Value {
+	if k == nil {
+		return nil
+	} else if v, ok := k.(graph.PreFetchedValue); ok {
+		return v.NameOf()
+	}
+	if v, ok := k.(Int64Value); ok {
+		if v == 0 {
+			if clog.V(2) {
+				clog.Infof("k was 0")
+			}
+			return nil
+		}
+		var val quad.Value
+		err := qs.db.View(func(tx *bolt.Tx) error {
+			var err error
+			val, err = qs.getValFromLog(tx, uint64(v))
+			return err
+		})
+		if err != nil {
+			clog.Errorf("error getting NameOf %d: %s", v, err)
+			return nil
+		}
+		return val
+	}
+	panic("unknown type of graph.Value; not meant for this quadstore. apparently a " + fmt.Sprintf("%#v", k))
+}
+
+func (qs *QuadStore) Quad(k graph.Value) quad.Quad {
+	var v quad.Quad
+	key := k.(*proto.Primitive)
+	err := qs.db.View(func(tx *bolt.Tx) error {
+		var err error
+		v, err = qs.primitiveToQuad(tx, key)
+		return err
+	})
+	if err != nil {
+		clog.Errorf("error fetching quad %d: %s", key, err)
+	}
+	return v
+}
+
+func (qs *QuadStore) primitiveToQuad(tx *bolt.Tx, p *proto.Primitive) (quad.Quad, error) {
+	q := &quad.Quad{}
```

Comment:
Right, but it can be just `var q quad.Quad`, and you can avoid `*q` in each return. Not a big deal though.